diff --git a/src/utilities.cpp b/src/utilities.cpp
index 50f7438..40d59bf 100644
--- a/src/utilities.cpp
+++ b/src/utilities.cpp
@@ -225,6 +225,7 @@ void setDefaultAppForType(const Fm::GAppInfoPtr app, std::shared_ptr<const Fm::M
 
 // templateFile is a file path used as a template of the new file.
 void createFileOrFolder(CreateFileType type, FilePath parentDir, const TemplateItem* templ, QWidget* parent) {
+    parent = nullptr;
     QString defaultNewName;
     QString prompt;
     QString dialogTitle = type == CreateNewFolder ? QObject::tr("Create Folder")
