diff --git a/src/plugin/wayfire-information.cpp b/src/plugin/wayfire-information.cpp
index 42a2d7a..8841544 100644
--- a/src/plugin/wayfire-information.cpp
+++ b/src/plugin/wayfire-information.cpp
@@ -87,12 +87,15 @@ void wayfire_information::send_view_info(wayfire_view view)
 
     pid_t pid = 0;
     wlr_surface *wlr_surface = view->get_wlr_surface();
+    wlr_xwayland_surface *xwayland_surface;
     int is_xwayland_surface = 0;
 #if WF_HAS_XWAYLAND
-    is_xwayland_surface = wlr_surface_is_xwayland_surface(wlr_surface);
-    if (is_xwayland_surface)
+    xwayland_surface = wlr_xwayland_surface_try_from_wlr_surface(wlr_surface);
+
+    if (xwayland_surface != NULL)
     {
-        pid = wlr_xwayland_surface_from_wlr_surface(wlr_surface)->pid;
+        is_xwayland_surface = 1;
+        pid = xwayland_surface->pid;
     } else
 #endif
     {
