diff --git a/arch/arm64/boot/dts/qcom/msm8916.dtsi b/arch/arm64/boot/dts/qcom/msm8916.dtsi
index cb85f77cd..c0d5ae571 100644
--- a/arch/arm64/boot/dts/qcom/msm8916.dtsi
+++ b/arch/arm64/boot/dts/qcom/msm8916.dtsi
@@ -258,6 +258,23 @@ opp-800000000 {
 		opp-998400000 {
 			opp-hz = /bits/ 64 <998400000>;
 		};
+		
+		// over clock-frequency
+		opp-1363200000 {
+		    opp-hz = /bits/ 64 <1363200000>;
+		};
+		opp-1401600000 {
+		    opp-hz = /bits/ 64 <1401600000>;
+		};
+		opp-1621600000 {
+		    opp-hz = /bits/ 64 <1621600000>;
+		};
+		opp-1841600000 {
+		    opp-hz = /bits/ 64 <1841600000>;
+		};
+		opp-1951600000 {
+		    opp-hz = /bits/ 64 <1951600000>;
+		};
 	};
 
 	firmware {
diff --git a/drivers/clk/qcom/a53-pll.c b/drivers/clk/qcom/a53-pll.c
index f43d455ab..9a3ccc58b 100644
--- a/drivers/clk/qcom/a53-pll.c
+++ b/drivers/clk/qcom/a53-pll.c
@@ -25,6 +25,9 @@ static const struct pll_freq_tbl a53pll_freq[] = {
 	{ 1248000000, 65, 0x0, 0x1, 0 },
 	{ 1363200000, 71, 0x0, 0x1, 0 },
 	{ 1401600000, 73, 0x0, 0x1, 0 },
+	{ 1621600000, 84, 0x0, 0x1, 0 },
+	{ 1841600000, 96, 0x0, 0x1, 0 },
+	{ 1951600000, 103, 0x0, 0x1, 0 },
 	{ }
 };
 
