diff --git a/plugins/common/wayfire/plugins/common/workspace-wall.hpp b/plugins/common/wayfire/plugins/common/workspace-wall.hpp
index 88042168..16d083de 100644
--- a/plugins/common/wayfire/plugins/common/workspace-wall.hpp
+++ b/plugins/common/wayfire/plugins/common/workspace-wall.hpp
@@ -390,19 +390,19 @@ class workspace_wall_t : public wf::signal::provider_t
                     self->wall->render_wall(target, region);
                 } else
                 {
-                    auto fb_region = target.framebuffer_region_from_geometry_region(region);
+                    //auto fb_region = target.framebuffer_region_from_geometry_region(region);
 
-                    OpenGL::render_begin(target);
-                    for (auto& dmg_rect : fb_region)
-                    {
-                        target.scissor(wlr_box_from_pixman_box(dmg_rect));
-                        const float a = 1.0 - dim;
+                    //OpenGL::render_begin(target);
+                    //for (auto& dmg_rect : fb_region)
+                    //{
+                    //    target.scissor(wlr_box_from_pixman_box(dmg_rect));
+                    //    const float a = 1.0 - dim;
 
-                        OpenGL::render_rectangle(target.geometry, {0, 0, 0, a},
-                            target.get_orthographic_projection());
-                    }
+                    //    OpenGL::render_rectangle(target.geometry, {0, 0, 0, a},
+                    //        target.get_orthographic_projection());
+                    //}
 
-                    OpenGL::render_end();
+                    //OpenGL::render_end();
                 }
             }
 
