diff --git a/src/output/render-manager.cpp b/src/output/render-manager.cpp
index a009ef43..b72da1b4 100644
--- a/src/output/render-manager.cpp
+++ b/src/output/render-manager.cpp
@@ -708,6 +708,20 @@ class wf::render_manager::impl
 
     wf::option_wrapper_t<wf::color_t> background_color_opt;
 
+    void paint_twice()
+    {
+        paint();
+
+        int cnt_acquired = 0;
+        for (int i = 0; i < WLR_SWAPCHAIN_CAP; i++) {
+            cnt_acquired += output->handle->swapchain->slots[i].acquired;
+        }
+
+        if (cnt_acquired <= 2) {
+            paint();
+        }
+    }
+
     impl(output_t *o) :
         output(o)
     {
@@ -726,14 +740,14 @@ class wf::render_manager::impl
             // https://github.com/swaywm/sway/pull/4588
             if (repaint_delay < 1)
             {
-                paint();
+                paint_twice();
             } else
             {
                 output->handle->frame_pending = true;
                 repaint_timer.set_timeout(repaint_delay, [=] ()
                 {
                     output->handle->frame_pending = false;
-                    paint();
+                    paint_twice();
                 });
             }
 
