From 2dd6b3776c46bebef704499d8899f5bb9a5ac7d5 Mon Sep 17 00:00:00 2001
From: hoream <2175223953@qq.com>
Date: Sat, 17 Jun 2023 23:56:10 +0800
Subject: [PATCH] patch

---
 wayfire/rule/lambda_rule.cpp | 1 +
 1 file changed, 1 insertion(+)

diff --git a/wayfire/rule/lambda_rule.cpp b/wayfire/rule/lambda_rule.cpp
index 99c558e..893d2bc 100644
--- a/wayfire/rule/lambda_rule.cpp
+++ b/wayfire/rule/lambda_rule.cpp
@@ -5,6 +5,7 @@
 #include <memory>
 #include <sstream>
 #include <string>
+#include <cstdint>
 
 namespace wf
 {
-- 
2.41.0

