From 068e59749f94d9d303117a5e955c233c31fcca07 Mon Sep 17 00:00:00 2001
From: Ilia Bozhinov <ammen99@gmail.com>
Date: Thu, 21 Mar 2024 17:50:08 +0200
Subject: [PATCH] output-layout: fix null pointer caused by #2229 (#2232)

---
 src/core/output-layout.cpp | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/src/core/output-layout.cpp b/src/core/output-layout.cpp
index d49d8ac5e..a0d4fbd20 100644
--- a/src/core/output-layout.cpp
+++ b/src/core/output-layout.cpp
@@ -10,6 +10,7 @@
 #include "../output/output-impl.hpp"
 #include <xf86drmMode.h>
 #include <cstring>
+#include <climits>
 #include <unordered_set>
 #include <drm_fourcc.h>
 #include <wayfire/seat.hpp>
@@ -67,7 +68,7 @@ wlr_output_mode *find_matching_mode(wlr_output *output,
                 return mode;
             }
 
-            const int bestSoFar = std::abs(best->refresh - reference.refresh);
+            const int bestSoFar = best ? std::abs(best->refresh - reference.refresh) : INT_MAX;
             const int current   = std::abs(mode->refresh - reference.refresh);
             if (!best || (bestSoFar > current))
             {
