diff --git a/wayland.c b/wayland.c
index f9ac162..3b8fe96 100644
--- a/wayland.c
+++ b/wayland.c
@@ -1298,8 +1298,9 @@ update_size(struct wayland *wayl)
     zwlr_layer_surface_v1_set_size(
         wayl->layer_surface, wayl->width / scale, wayl->height / scale);
 
-    zwlr_layer_surface_v1_set_anchor(
-        wayl->layer_surface, wayl->conf->anchor);
+    if (wayl->conf->anchor != 15)
+        zwlr_layer_surface_v1_set_anchor(
+            wayl->layer_surface, wayl->conf->anchor);
 
     /* Trigger a 'configure' event, after which we'll have the actual width+height */
     wl_surface_commit(wayl->surface);
