diff --git a/src/decoder/Thread.cxx b/src/decoder/Thread.cxx
index 60ff299f5..13abfb85d 100644
--- a/src/decoder/Thread.cxx
+++ b/src/decoder/Thread.cxx
@@ -212,8 +212,7 @@ decoder_run_stream_plugin(DecoderBridge &bridge, InputStream &is,
 			  std::string_view suffix,
 			  const DecoderPlugin &plugin)
 {
-	if (!decoder_check_plugin_mime(plugin, is) &&
-	    !decoder_check_plugin_suffix(plugin, suffix))
+	if (!decoder_check_plugin_suffix(plugin, suffix))
 		return DecodeResult::NO_PLUGIN;
 
 	if (plugin.stream_decode == nullptr)
